### PR TITLE
Added SMART filesystem to RP2040

### PR DESCRIPTION
**Pull request by curuvar**
## Summary
 
This code implements a Sector Mapped Allocation for Really Tiny (SMART) filesystem in the RP2040 flash memory chip.  It uses the space not otherwise used by the NuttX binary and supports both read and write access.  The initial state if the filesystem can be set when a new program is uploaded to the RP2040.  Data subsequently written to the filesystem persists across re-boots.

## Impact

This bulk of this request is RP2040 only, although a few common SMART filesystem source files were updated to correctly handle unaligned access.

## Testing

This change has been tested on a RaspberryPi pico W. Data has been read and written by various NSH commands and the tests in smart_test work as expected.

### CODE REVIEW

File: fs/smartfs/smartfs_smart.c [hunk 1]
```diff
@@ -657,6 +675,18 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer,
 
   DEBUGASSERT(filep->f_priv != NULL && filep->f_inode != NULL);
 
+#ifdef CONFIG_DEBUG_FS_INFO
+  finfo("Writing %lu bytes:", (uint32_t)buflen);
+  for (uint32_t i = 0; i < buflen; ++i)
+    {
+      if ((i & 0x0f) == 0) printf("\n    ");
+
+      printf("%02x, ", buffer[i]);
```
Comment (xiaoxiang781216) [inline fs/smartfs/smartfs_smart.c:684]: kernel shouldn't call printf

File: fs/smartfs/smartfs_smart.c [hunk 2]
```diff
@@ -633,6 +639,18 @@ static ssize_t smartfs_read(FAR struct file *filep, char *buffer,
 
   /* Return the number of bytes we read */
 
+#ifdef CONFIG_DEBUG_FS_INFO
+  finfo("Read %lu bytes:", bytesread);
+  for (uint32_t i = 0; i < bytesread; ++i)
+    {
+      if ((i & 0x0f) == 0) printf("\n    ");
+
+      printf("%02x, ", buffer[i]);
```
Comment (xiaoxiang781216) [inline fs/smartfs/smartfs_smart.c:648]: ditto

File: fs/smartfs/smartfs_smart.c [hunk 3]
```diff
@@ -528,6 +532,8 @@ static ssize_t smartfs_read(FAR struct file *filep, char *buffer,
 
   DEBUGASSERT(filep->f_priv != NULL && filep->f_inode != NULL);
 
+  finfo("Reading %u bytes\n", (unsigned) buflen);
```
Comment (xiaoxiang781216) [inline fs/smartfs/smartfs_smart.c:535]:
```
finfo("Reading %zu bytes\n", buflen);
```

File: fs/smartfs/smartfs_smart.c [hunk 4]
```diff
@@ -941,6 +971,8 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer,
 
   ret = byteswritten;
 
+  finfo("Wrote %u bytes\n", (unsigned) byteswritten);
```
Comment (xiaoxiang781216) [inline fs/smartfs/smartfs_smart.c:974]:
```
finfo("Reading %zu bytes\n", byteswritten);
```

File: fs/smartfs/smartfs_smart.c [hunk 5]
```diff
@@ -45,6 +45,10 @@
 
 #include "smartfs.h"
 
+#ifdef CONFIG_DEBUG_FS_INFO
+#  include <stdio.h>
```
Comment (xiaoxiang781216) [inline fs/smartfs/smartfs_smart.c:49]: remove, you can't call printf inside kernel

File: fs/smartfs/smartfs_smart.c [hunk 6]
```diff
@@ -852,7 +882,7 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer,
           /* Copy the new sector to the old one and chain it */
 
           header = (struct smartfs_chain_header_s *) sf->buffer;
-          *((uint16_t *) header->nextsector) = (uint16_t) ret;
+          SMARTFS_SET_NEXTSECTOR(header, (uint16_t) ret);
```
Comment (xiaoxiang781216) [inline fs/smartfs/smartfs_smart.c:885]: let's move all cast to SMARTFS_SET_NEXTSECTOR macro

**Comment by masayuki2009** [comment]
The following issue has been reported.
https://github.com/apache/incubator-nuttx/issues/7233

**Comment by masayuki2009** [comment]
I think this PR should have been separated into several commits.

**Pull request merged**